Add application insights connection string support - APPLICATIONINSIGHTS_CONNECTION_STRING

Greybird opened the pull request:
**Task names**: 
- AzureFunctionAppContainerV1
- AzureFunctionAppV1
- AzureFunctionAppV2
- AzureRmWebAppDeploymentV3
- AzureRmWebAppDeploymentV4
- AzureWebAppContainerV1
- AzureWebAppV1

**Description**:
The PR adds support for adding release annotations when the now recommended APPLICATIONINSIGHTS_CONNECTION_STRING setting is used on Azure, rather than the APPINSIGHTS_INSTRUMENTATIONKEY, for all tasks supporting this feature.

**Documentation changes required:**
I don't think so.

**Added unit tests:**
No.
I did not find tests for release annotation based on InstrumentationKey.
I had a hard time figuring how to add unit tests for the connection string use case, and tests seem to come from another zone of the code.
Happy to get guidance on the matter however.

**Attached related issue:**
#18796 (closed due to being stale, probably needs reopening)

**Checklist**:
- [X] Task version was bumped - please check [instruction](https://github.com/microsoft/azure-pipelines-tasks/tree/master/docs/taskversionbumping.md) how to do it
- [X] Checked that applied changes work as expected (tested on our organization by publishing a preview task as explained in documentation)

FinVamp1 commented on Tasks/AzureFunctionAppV1/operations/ReleaseAnnotationUtility.ts:
We're making newer changes to the V2 task for Azure Functions Apps rather than in the v1 release. Do you view this as a  potential issuie if this is only supported for Connection String Release annotation in the v2 release?

Greybird replied:
Hi @FinVamp1, thanks for the feedback.
I don't see this as an issue at all, I chose to update all tasks where the instrumentation key behavior was present, but as far as the latest version has the connection string support, it is fine to me.
Tell me if you want me to revert the changes I made for AzureFunctionAppV1 (and potentially AzureRmWebAppDeploymentV3 too if pertinent).

FinVamp1 replied:
HI, I think the other aspect to is if it we don't have this change in down-level tasks then we could break the release annotations if customers migrate to the Connection String.  I think this change is low level enough that we could have it in both versions. Are you also going to update the Task.json and the Task Version to ensure your change is merged? https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/taskversionbumping.md

Greybird replied:
TBH, it already broke, as connection strings are required for new regions, and migration is promoted since a few time already. From the issue linked (#18796, created august 2023) and [this comment on issue 14961](https://github.com/microsoft/azure-pipelines-tasks/issues/14961#issuecomment-1322620524) (dated Nov 2022)
My contribution is also linked to our company loosing these annotations in the process of migrating.

FinVamp1 replied:
Agreed, and I've approved it from the Functions side of things.

FinVamp1 reviewed (changes requested):
Can you update the Task.json as mentioned?

Greybird commented:
> Can you update the Task.json as mentioned?

Sure, I just did, and updated issue template too. Feel free to point any mistake I would have made in the process of course!

FinVamp1 reviewed (approved):
Looks good to me from the Function side of things. Thanks for the contribution.

jvano reviewed (approved):
:shipit:

Greybird commented:
Thanks @jvano, @manolerazvan !

@dannysongg, @patelchandni, could you please take a look ?

FinVamp1 commented:
> Thanks @jvano, @manolerazvan !
> 
> @dannysongg, @patelchandni, could you please take a look ?

Hello @Greybird I think in this instance Joaquin and Danny are part of the Web Apps task team and Chandni and I are part of the Functions team so just one approval is probably needed from each team.

Greybird commented:
@FinVamp1, you are right, sorry,
I have been confused by the "2 pending reviewers" mention, while the changes are actually approved. Thanks for pointing this!

I suppose there is nothing left for me to do currently, as checks now have to run, which I believe depends on some validation on Microsoft side ?
Do not hesitate to point to needed actions if I am wrong.

Thanks again for your guidance with the process !

Greybird commented:
Hello,

I updated sprint to 241.
I don't know how I can help further to have this PR merged.

Greybird commented:
@v-schhabra sorry to ping you, but I saw a rebase on master from you, can you provide guidance ? Every option to make the PR more mergeable in the documentation is not possible for me (cannot label, cannot add a reviewer, and I already linked an issue, but I cannot add it to the PR.)

v-schhabra commented:
> @v-schhabra sorry to ping you, but I saw a rebase on master from you, can you provide guidance ? Every option to make the PR more mergeable in the documentation is not possible for me (cannot label, cannot add a reviewer, and I already linked an issue, but I cannot add it to the PR.)

Hi @Greybird 
Sorry for replying late somehow I missed your message.
Do you have write permissions for this repo? If no then Ig you will not have the option to merge the PR.

v-schhabra commented:
/azp run

Greybird commented:
@v-schhabra

> Do you have write permissions for this repo? If no then Ig you will not have the option to merge the PR.
> 

No I don't have write permissions. But I'm mentionning instructions from the documentation for contributors here that ask for actions that do not seem to be doable for non Microsoft contributors:
https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/contribute.md#pr-submission

Thanks for launching checks. I see bad credentials errors, I suppose this is out of my reach to fix this?

v-schhabra commented:
@Greybird You are getting bad credential error becoz you didnt have permissions.

Greybird commented:
> @Greybird You are getting bad credential error becoz you didnt have permissions for this repo.

On the check pipelines?
How could the PR be validated then?

v-schhabra commented:
> > @Greybird You are getting bad credential error becoz you didnt have permissions for this repo.
> 
> On the check pipelines? How could the PR be validated then?

I will check with my team on how can we proceed further here and will keep you posted on the updates.

v-schhabra commented:
Hi @Greybird 
I could see you have not updated task.json file for AzureFunctionAppContainerV1, AzureFunctionAppV1, AzureFunctionAppV2, AzureWebAppContainerV1 and AzureWebAppV1 tasks. 
Incase you missed to update task.json file please update it or is it purposefully ignored?

Greybird commented:
Hi @v-schhabra ,

Thanks for pointing this. Indeed the merge with master collided with changes, and I forgot to apply versioning to these task.json files.
I merged again with master, and reapplied patch versions.
To me everything should be correct now, but if you find anything missing, please tell me!

Greybird commented:
Forgot to answer your question @v-schhabra: yes I conducted tests by publishing these tasks under new guids to a private org, and they worked.

Greybird commented:
Hi @v-schhabra,
I rebased on master again, ans switched to sprint 243.
Did you find a way to have pipelines pass on the PR ? Is there another way to manage this situation ?

Greybird commented:
Hi @v-schhabra ,
I rebased on master (again), and bumped the version.

As this PR fixes a Microsoft-introduced issue following a change of recommendation regarding how Application Insights should be declared on infrastructure, I would expect a more straightforward road to a merge, especially as code owners validated the content already.
I must admit that the contribution experience is rather frustrating, and not in line with how I undestand the spitit of [contribute.md](https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/contribute.md), with 2 and a half month of bumping sprint numbers, patch versions, and integrating master changes.

If not being part of microsoft is what makes it impossible to have the build pipelines pass, then is there a way to have someone from Microsoft (you?) create another PR with the same content to make this PR mergeable ?

Thanks for the input on this matter,

FinVamp1 commented:
@Greybird I've merged your changes into the Functions Apps V2 task as part of this PR for other work. https://github.com/microsoft/azure-pipelines-tasks/pull/20191 

@v-schhabra  and @manolerazvan  can you help Arnaud get the rest of his changes in? This is an important community contribution I think that will help as indicated above.

v-schhabra commented:
> Hi @v-schhabra , I rebased on master (again), and bumped the version.
> 
> As this PR fixes a Microsoft-introduced issue following a change of recommendation regarding how Application Insights should be declared on infrastructure, I would expect a more straightforward road to a merge, especially as code owners validated the content already. I must admit that the contribution experience is rather frustrating, and not in line with how I undestand the spitit of [contribute.md](https://github.com/microsoft/azure-pipelines-tasks/blob/master/docs/contribute.md), with 2 and a half month of bumping sprint numbers, patch versions, and integrating master changes.
> 
> If not being part of microsoft is what makes it impossible to have the build pipelines pass, then is there a way to have someone from Microsoft (you?) create another PR with the same content to make this PR mergeable ?
> 
> Thanks for the input on this matter,

Apologies for replying late.
Recently I have some high priority items in my queue once I have some bandwidth will create another PR by having these changes. 
Will test the tasks and merge it then.
Thanks for the contribution and patience.

manolerazvan reviewed (commented):
can we ensure unit tests are covering both the old and the new scenarios ? especially different formats of APPLICATIONINSIGHTS_CONNECTION_STRING ?

v-schhabra commented:
> can we ensure unit tests are covering both the old and the new scenarios ? especially different formats of APPLICATIONINSIGHTS_CONNECTION_STRING ?

Sure, will check Razvan.

Greybird commented:
Hi @v-schhabra , any news ?

I rebased on master, and updated versions agaiin.
I removed AzureFunctionAppV1 from the commit as AzureFunctionAppV2 was already updated thanks to @FinVamp1.

This leaves:
- AzureFunctionAppContainerV1
- AzureRmWebAppDeploymentV3
- AzureRmWebAppDeploymentV4
- AzureWebAppContainerV1
- AzureWebAppV1

to update.

v-schhabra commented:
Hi @Greybird 
I did the changes on my end for the tasks owned by our team. Will test the changes and create the PR soon.

Greybird commented:
Great news, thanks for your work on the subject!

manolerazvan commented:
@dannysongg @jvano  these leaves us with the following

AzureFunctionAppContainerV1
AzureRmWebAppDeploymentV3
AzureRmWebAppDeploymentV4

Are you able to test the changes and merge them ?

dannysongg commented:
@Greybird I'll be updating AzureRmWebAppDeploymentV3 and AzureRmWebAppDeploymentV4 on the WebApp side. 
@FinVamp1 do you own AzureFunctionAppContainerV1?
